lib/streamlit/state/session_state.py

```diff
@@ -278,6 +304,62 @@ def _missing_key_error_message(key: str) -> str:
     )
 
 
+class WidgetStateReport(Protocol[T_co]):
+    """Should be usable by widget code to determine what value to return, and
+    whether to update the UI.
+    """
+
+    @property
+    @abstractmethod
+    def return_value(self) -> T_co:
+        ...
+
+    @property
+    @abstractmethod
+    def set_frontend_value(self) -> bool:
+        ...
+
+
+@attr.s(auto_attribs=True, frozen=True, slots=True)
+class RealWidgetState(WidgetStateReport[T_co]):
+    """A WidgetStateReport signifying that an actual value was retrieved from
+    SessionState.
+    """
+
+    value: T_co
+    value_changed: bool
+
+    @property
+    def return_value(self) -> T_co:
+        return self.value
+
+    @property
+    def set_frontend_value(self) -> bool:
+        return self.value_changed
+
+
+@attr.s(auto_attribs=True, frozen=True, slots=True)
+class FallbackState(WidgetStateReport[T]):
+    """A WidgetStateReport signifying that SessionState could not be accessed.
+
+    Provides a fallback return_value, and suggests not updating the UI.
+    """
+
+    deserializer: WidgetDeserializer[T]
+
+    @classmethod
+    def from_metadata(cls, metadata: WidgetMetadata[T]) -> "FallbackState[T]":
+        return cls(deserializer=metadata.deserializer)
+
+    @property
+    def return_value(self) -> T:
+        return self.deserializer(None, "")
+
+    @property
+    def set_frontend_value(self) -> Literal[False]:
+        return False
```

Comment:
I think the protocol + two concrete implementations adds unnecessary conceptual complexity (and a lot of code). Callers of `register_widget` don't care if they receive a "FallbackState" or a "RealWidgetState", so their existence doesn't clarify the _API_. And since both classes are just tuples storing two values, I think their existence also doesn't make the `register_widget` implementation any simpler.

I think this could all be simplified to something like this (docstrings omitted):

```python
@attr.s(auto_attribs=True, frozen=True, slots=True)
class RegisterWidgetResult(Generic[T_co]):
    value: T_co
    value_changed: bool
```

(Also: let's not have property names be different from the variable names they wrap - just feels like it adds code and complexity.)

If it's important that a "failed" widget registration be easily constructed, this could be expanded to:

```python
@attr.s(auto_attribs=True, frozen=True, slots=True)
class RegisterWidgetResult(Generic[T_co]):
    @classmethod
    def failure(cls, metadata: WidgetMetadata[T]) -> "RegisterWidgetResult[T]":
        return cls(value=metadata.deserializer(None, ""), value_changed=False)

    value: T_co
    value_changed: bool
```

Reply:
I went with your solution. Some minor comments:

> Callers of register_widget don't care if they receive a "FallbackState" or a "RealWidgetState"

The caller doesn't, but the developer might. I realize it`s a bit overkill to partition the type like this for documentation purposes, though.

> If it's important that a "failed" widget registration be constructed in the class itself...

I think it serves a purpose in terms of self-documentaion. It provides a single, canonical way of creating the result in failure cases, which seems to be needed, given that I had to fix an exception to this as part of the PR.